### PR TITLE
add async or sync for :entry location index's rocksdb write

### PR DESCRIPTION
**Pull request by StevenLuMT**
Descriptions of the changes in this PR:

### Motivation

add async mode for :entry location index's rocksdb write: 
1)  default sync mode: not change the original logic
2)  async mode : need update config to open, this mode is to speed up writing

### Changes

1. add async or sync for :location rocksdb write
2. add switch to open this feature, default not open, use default sync

### CODE REVIEW

**Comment by StevenLuMT** [comment]
rerun failure checks

**Comment by StevenLuMT** [comment]
@eolivelli @pkumar-singh @zymap
If you have time, please help me review it, thank you

**Comment by StevenLuMT** [comment]
@eolivelli @pkumar-singh @zymap
If you have time, please help me review it, thank you

**Comment by merlimat** [comment]
@StevenLuMT can you add in the motivation section the reasons for this change and what are the trade offs here?

**Comment by StevenLuMT** [comment]
> @StevenLuMT can you add in the motivation section the reasons for this change and what are the trade offs here?
@merlimat 
async mode :increase write rate
default sync mode: not change the original logic

**Comment by StevenLuMT** [comment]
@eolivelli @pkumar-singh @zymap @hangc0276 
If you have time, please help me review it, thank you.

**Comment by StevenLuMT** [comment]
@eolivelli @pkumar-singh @zymap @hangc0276
If you have time, please help me review it, thank you.

**Comment by StevenLuMT** [comment]
@dlg99 @eolivelli @pkumar-singh @zymap @hangc0276
If you have time, please help me review it, thank you.

**Comment by StevenLuMT** [comment]
@dlg99 @eolivelli @pkumar-singh @zymap @hangc0276
If you have time, please help me review it, thank you.

**Comment by StevenLuMT** [comment]
@dlg99 @eolivelli @pkumar-singh @zymap @hangc0276
If you have time, please help me review it, thank you.

**Comment by StevenLuMT** [comment]
@dlg99 @eolivelli @pkumar-singh @zymap @hangc0276
If you have time, please help me review it, thank you.

**Comment by StevenLuMT** [comment]
@dlg99 @eolivelli @pkumar-singh @zymap @hangc0276 @sijie 
If you have time, please help me review it, thank you.

**Comment by StevenLuMT** [comment]
@dlg99 @eolivelli @pkumar-singh @zymap @hangc0276 @lordcheng10 
If you have time, please help me review it, thank you.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/EntryLocationIndexAsync.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        }
+
+        // Update all the ledger index pages with the new locations
+        for (EntryLocation e : newLocations) {
+            if (log.isDebugEnabled()) {
+                log.debug("Update location - ledger: {} -- entry: {}", e.ledger, e.entry);
+            }
+            addLocation(null, e.ledger, e.entry, e.location);
+        }
+        locationsDb.sync();
```
Comment (lordcheng10) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/EntryLocationIndexAsync.java:72]: Here why use locationsDb.sync(); to replace the following logic: 
batch.flush();
batch.close();

Comment (lordcheng10) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/EntryLocationIndexAsync.java:72]: What is the difference between locationsDb.sync(); and batch.flush();?
In the end it is called db.write(optionSync, emptyBatch);
image
image

It doesn't look asynchronous here.

Comment (StevenLuMT) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/EntryLocationIndexAsync.java:72]: sync is different from async,

- sync mode:  
      1.  create a batch; 
      2.  add msg to the batch
      3.  call method(batch.flush) to flush the batch

- sync mode:  
      1. just call method(locationsDb.sync)  to write the data
      2. the rocksdb server will be timed to flush the data

**Pull request closed without merging**